starlingcaptureapi/actions.py

```diff
@@ -22,18 +23,102 @@
+            organization_id: ID of the organization
+            collection_id: ID of the collection this asset is under
+            zip_path: path to the ZIP that contains the asset and metadata
+            asset_exts: valid file extensions for the asset (no dots)
+            aes_key: 32 byte key for AES-256-CBC encryption
+
+        Returns:
+            TODO
 
         Raises:
             Exception if errors are encountered during processing
         """
-        archive = EncryptedArchive.make_from_meta(asset_meta_path)
-        Iscn.register_archive(archive)
 
+        asset_helper = AssetHelper(organization_id)
+        file_util = FileUtil()
+
+        tmp_dir = asset_helper.get_tmp_collection_dir(collection_id, "archive")
+
+        # Copy ZIP
+        tmp_zip = shutil.copy2(zip_path, tmp_dir)
+
+        # Verify zip contents are valid and expected
+
+        zip_listing = zip_util.listing(tmp_zip)
+        if len(zip_listing) > 3:
+            # Should only have three: content, recorder meta, content meta
+            raise Exception(
+                f"ZIP at {zip_path} has more than three files: {zip_listing}"
+            )
+
+        content_filename = next((s for s in zip_listing if "-" not in s), None)
+        if content_filename is None:
+            raise Exception(f"ZIP at {zip_path} has no content file: {zip_listing}")
+
+        if "/" in content_filename:
+            raise Exception(f"Content file is not at ZIP root: {content_filename}")
+
+        if os.path.splitext(content_filename)[1][1:] not in asset_exts:
+            raise Exception(
+                f"Content file in ZIP has wrong extension: {content_filename}"
+            )
+
+        # Verify hash
+
+        content_sha = os.path.splitext(os.path.basename(zip_path))[0]
+        zip_dir = os.path.join(tmp_dir, content_sha)
+        extracted_content = os.path.join(zip_dir, content_filename)
+
+        file_util.create_dir(zip_dir)
+        zip_util.extract_file(tmp_zip, content_filename, extracted_content)
+
+        if content_sha != file_util.digest_sha256(extracted_content):
+            raise Exception("SHA-256 hash of content file does not match ZIP name")
+
+        # Generate other hashes
+
+        content_cid = file_util.digest_cidv1(extracted_content)
+        content_md5 = file_util.digest_md5(extracted_content)
+
+        # Register on OpenTimestamp and add that file to zip
+
+        content_ots = extracted_content + ".ots"
+        file_util.register_timestamp(extracted_content, content_ots)
+        zip_util.append(
+            tmp_zip, content_ots, os.path.join("proofs", os.path.basename(content_ots))
+        )
+
+        # Get final ZIP hashes
+
+        zip_sha = file_util.digest_sha256(tmp_zip)
+        zip_md5 = file_util.digest_md5(tmp_zip)
+        zip_cid = file_util.digest_cidv1(tmp_zip)
+
+        # Move ZIP to storage
+        # TODO: name is ZIP hash, not asset hash right?
```

Comment:
Yea generally the asset refers to the content file (w/o metadata).

Reply:
So the final unencrypted ZIP should be named as the SHA-256 of itself. What about the encrypted ZIP? Is it the hash of itself, or the hash of the unencrypted ZIP?